104 - Basic AI Tile Visibility

QuintillusCFC opened the pull request:
Changes:

- AI stores info about which tiles it knows (does not yet import this from Civ3)
- Explorer AI now seeks to explore new tiles, rather than moving randomly

This doesn't accomplish all of the goals of #104 , in addition to not importing Civ3 data, the settler AI doesn't use tile knowledge, and the AI is a little too persistent about discovering tiles that have opposing cities on them.  But at 200 lines net, I want to at least have a draft PR in case adding more pushes it too high, and will likely split those out.

QuintillusCFC committed:
The AI will now keep track of which tiles it knows about.  This is updated whenever a unit moves.

Still a long ways to go, this is just the structure.

QuintillusCFC committed:
Explorer AI now prioritizes tiles that will reveal more tile knowledge.

At the beginning, this looks great, as AI unit fan out to discover new tiles.

But at some point the flaw - that it's always looking for the local maximum - reveals itself, as units get trapped in corners of the map.

Probably going to refactor things a bit before fixing that.

QuintillusCFC committed:
Yank the Explorer AI out into its own class.

QuintillusCFC committed:
WIP towards teaching the AI to choose a new exploration destination if it finds that it has already explored all the tiles immediately around an explorer unit.

QuintillusCFC committed:
Add the ExplorerAIData class that I appear to have forgotten to commit from the desktop.  Probably indicates the .gitignore needs to be added it, as usually this happens due to files hiding among other uncommitted files.

QuintillusCFC committed:
In the event an Explorer has no tiles right beside it to explore, it will now choose a path to the nearest tile that will let it keep exploring.

QuintillusCFC committed:
Units will now follow a path to a new area to explore when they've finished exploring their immediate area.

They still get stuck after they run out of areas on their continent, but that is expected at this point.

QuintillusCFC committed:
Restructure code to have much better, more understandable, and higher-level structure.

QuintillusCFC committed:
Fix conditional that caused AI to not seek more places to explore.

Now that the AI is exploring with a purpose, it might be neglecting defenses a bit too much.  All three AIs got destroyed by barbs (or in one case, an exploring AI) in my play-through.

QuintillusCFC committed:
Add a couple checks to reduce how often scouts will inadvertently attack cities.

pcen commented on C7Engine/AI/ExplorerAI.cs:
```diff
@@ -0,0 +1,131 @@
+
+namespace C7Engine
+{
+	public class ExplorerAI
```
It could be nice to have an interface for both AI and for AI data. Partially because it would be nice to treat AI polymorphically, but also because the current naming convention is slightly confusing for someone who isn't familiar with the code ie. we have `ExplorerAIData : UnitAI`, but I don't really expect implementations of `UnitAI` to contain data

QuintillusCFC replied:
Yeah, I was thinking about that when I fell asleep last night and forgot about it when I woke up.  UnitAI is definitely confusing, I will rename that.  I should be able to get the AI to be polymorphic as well.

QuintillusCFC replied:
I wound up spinning this (other than the UnitAI data rename) off into https://github.com/C7-Game/Prototype/pull/203, since it's a semi-decent chunk of changes and isn't really related to tile visibility/the initial explorer AI.  It is a nice improvement though, glad you mentioned it.

QuintillusCFC committed:
Rename UnitAI in C7GameData to UnitAIData because two interfaces with the same name is confusing.

Rename currentAIBehavior variable in MapUnit.cs to currentAIData, to be more consistent in showing that what's stored is the AI data, not the behavior classes.

pcen commented on C7GameData/AIData/TileKnowledge.cs:
```diff
@@ -0,0 +1,34 @@
+		 * Returns a copy of the list of known tiles.
+		 * This prevents external modifications.
+		 **/
+		public List<Tile> AllKnownTiles()
```
is there a particular reason for returning the copy as a list?

QuintillusCFC replied:
Potentially excessive caution around the data being modified outside of the TileKnowledge API.

The gist of the thinking was that there are places where it's handy for the AI to have a list of all known tiles - for example, deciding where to send a settler.  But it should never be possible to remove items from that list, and if you send the whole list out, it is.

In Java, I would have done this via the [Collections.unmodifiableList](https://docs.oracle.com/javase/1.5.0/docs/api/java/util/Collections.html#unmodifiableList(java.util.List)) method, which returns a read-only view of the list.  This is lighter weight than making a whole copy, but I'm not aware of an equivalent in C#.

pcen replied:
Yeah it makes senes to to return a copy, I was just curious if there's a reason why the "real" known tiles is in a map, but the AllKnownTiles returns a copy as a list, as opposed to a copy as a map.

QuintillusCFC replied:
Oh, right.

The internal collection is a Set so that if a tile is added more than once via _AddTilesToKnown_, it still only winds up there once.  My intention with _AddTilesToKnown_ is that by passing it the tile a unit is on, it can automatically handle logic around how many neighboring tiles should be added, including based on terrain height (hills, mountains, etc.).  A set seemed like the simplest way to avoid adding duplicates.

For consumption purposes, a list just seemed like the default collection.  A set would work just as well, though.

I'm guessing you meant "set" instead of "map" in both cases?

QuintillusCFC replied:
_Long_ term, we'll also need more methods to deal with tiles revealed by air missions... but that can probably just be a new method similar to AddTilesToKnown, but that doesn't take into account ground terrain, and has a wider radius.  I think it will be compatible, but am not too concerned about it now.

pcen reviewed (approved):
Since this is in conjunction with https://github.com/C7-Game/Prototype/pull/203, I'm just reviewing the AI logic, which looks good to me. I like how there's a comment for all the "temporary" logic, it will make it easy to iterate over in the future
